mlflow/server/handlers.py

```diff
@@ -1173,37 +1184,49 @@ def get_metric_history_bulk_interval_handler():
 
     store = _get_tracking_store()
 
-    def _get_max_step_for_metric(run_id, metric_key):
-        if hasattr(store, "get_max_step_for_metric"):
-            return store.get_max_step_for_metric(run_id=run_id, metric_key=metric_key)
-        steps = [m.step for m in store.get_metric_history(run_id, metric_key)]
-        return max(steps) if steps else 0
-
     def _get_sampled_steps(run_ids, metric_key, max_results):
         # cannot fetch from request_message as the default value is 0
         start_step = args.get("start_step")
         end_step = args.get("end_step")
+
+        # get a list of all steps for all runs. this is necessary
+        # because we can't assume that every step was logged, so
+        # sampling needs to be done on the steps that actually exist
+        all_runs = [
+            [m.step for m in store.get_metric_history(run_id, metric_key)] for run_id in run_ids
+        ]
+
+        # save mins and maxes to be added back later
+        all_mins_and_maxes = {step for run in all_runs if run for step in [min(run), max(run)]}
+
+        all_steps = sorted({step for sublist in all_runs for step in sublist})
```

Comment:
because we have to retrieve the min and max for each run, i'm not sure if there's a good way to condense this further.

in databricks, the retrieval of the min and max for each run is offloaded to the DB, but we don't necessarily have such functions here.

if we didn't have to get the min and max, then we could do something like:

```python
all_steps = sorted(
    {
        step
        for sublist in (
            [m.step for m in store.get_metric_history(run_id, metric_key)]
            for run_id in run_ids
        )
        for step in sublist
    }
)
```

Reply:
mostly using list and set comprehensions here because it seems they're optimized in python